types/wlr_data_control_v1.c

```diff
@@ -106,17 +82,134 @@ static const struct zwlr_data_control_source_v1_interface source_impl = {
 };
 
 static void source_handle_resource_destroy(struct wl_resource *resource) {
-	struct client_data_source *source = source_from_resource(resource);
+	struct data_control_source *source = source_from_resource(resource);
 	if (source == NULL) {
 		return;
 	}
-	wlr_data_source_destroy(&source->source);
+
+	char **p;
```

Comment:
Can we move all of this into a `data_control_source_destroy` function? This would allow to call it from here and other places without using `wl_resource`, which is not type-safe.

Reply:
Other places would still need to destroy the resource though which ends up running this code anyway. What would extracting it achieve?

Reply:
No, you should never destroy resources yourself. `wl_resource_destroy` should only be called in the destroy request handler.

Extracting it would make sure a completely difference resource isn't passed.

Reply:
> using `wl_resource`, which is not type-safe
(Completely unrelated to this pull request, but) do you happen to know why wayland-server is designed this way, especially given how wayland-client is typesafe (with all objects still being generic wl_proxies under the hood)?

Reply:
It seems to be historical raisins, according to Pekka. No real reason.

Reply:
> Can we move all of this into a `data_control_source_destroy` function?

Something like this?